tests/model_fields/test_imagefield.py

```diff
@@ -177,13 +177,14 @@ def test_pickle(self):
         p = Person(name="Joe")
         p.mugshot.save("mug", self.file1)
         dump = pickle.dumps(p)
-
-        p2 = Person(name="Bob")
-        p2.mugshot = self.file1
+        mugshot_dump = pickle.dumps(p.mugshot)
 
         loaded_p = pickle.loads(dump)
         self.assertEqual(p.mugshot, loaded_p.mugshot)
 
+        loaded_mugshot = pickle.loads(mugshot_dump)
+        self.assertEqual(p.mugshot.url, loaded_mugshot.url)
```

Comment:
I'm not sure if this is a good solution, because `p.mugshot.storage != loaded_mugshot.storage` and in the same time `loaded_p.mugshot.storage == p.mugshot.storage` :thinking:

Reply:
Yes, you are right.
Here is my understanding of the problem. 
Based on [python pickle documentation ](https://docs.python.org/3/library/pickle.html#pickling-class-instances), when we `unpickle` an object it creates an uninitialized instance and then restores the saved attributes.
So, when we `pickle` and `unpickle` an `ImageField` by:
```
mugshot_dump = pickle.dumps(p.mugshot)
loaded_mugshot = pickle.loads(mugshot_dump)
```
we will have:
```
p.mugshot.storage != loaded_mugshot.storage
```

We should have the same behavior when we `pickle` and `unpickle` an `Person` by:
```
dump = pickle.dumps(p)
loaded_p = pickle.loads(dump)
```
But, in this case, we will have:
```
p.mugshot.storage == loaded_p.mugshot.storage
```
because of this line:
https://github.com/django/django/blob/6e2f05b2e33a6c80c7a411ce76af7b5a08acb835/django/db/models/fields/files.py#L199

Reply:
I think that this issue is more complicated and a lack of `storage` is only a part of the problem. `instance` and `field` are also unavailable and other attributes may not be set correctly. `FieldField` (and subclasses) cannot be (at least currently) pickled and unpickled because `FileDescriptor` is responsible for setting/restoring most of attributes and it is not involved in this scenario. My understanding is that we should provide `__setstate__()` (see [documentation](https://docs.python.org/3/library/pickle.html#handling-stateful-objects)), e.g. (just an idea)
```python
    def __getstate__(self):
        # ....
        return {
            'name': self.name,
            'closed': False,
            '_committed': True,
            '_file': None,
            'instance': self.instance,
            'field': self.field,
        }

    def __setstate__(self, state):
        self.__dict__.update(state)
        self.storage = self.field.storage
```